be/src/util/json.cpp

```diff
@@ -188,31 +205,44 @@ static int sliceCompare(const vpack::Slice& left, const vpack::Slice& right) {
             idx++;
         }
         return 0;
-    } else if (left.type() == right.type()) {
-        auto type = left.type();
-        switch (type) {
-        case vpack::ValueType::Null:
-            return 0;
-        case vpack::ValueType::Bool:
-            return left.getBool() - right.getBool();
-        case vpack::ValueType::SmallInt:
-        case vpack::ValueType::Int:
-        case vpack::ValueType::UInt:
-            return left.getIntUnchecked() - right.getIntUnchecked();
-        case vpack::ValueType::Double: {
-            if (std::isless(left.getDouble(), right.getDouble())) {
-                return -1;
-            } else if (std::isgreater(left.getDouble(), right.getDouble())) {
-                return 1;
+    } else if (vpack::valueTypeGroup(left.type()) == vpack::valueTypeGroup(right.type())) {
+        // 1. type are exactly same
+        // 2. type are both number, but could smallInt/Int/Double
+        if (left.type() == right.type()) {
+            switch (left.type()) {
+            case vpack::ValueType::Bool:
+                return left.getBool() - right.getBool();
+            case vpack::ValueType::SmallInt:
+            case vpack::ValueType::Int:
+            case vpack::ValueType::UInt:
+                return left.getInt() - right.getInt();
+            case vpack::ValueType::Double: {
+                return cmpDouble(left.getDouble(), right.getDouble());
             }
-            return 0;
-        }
-        case vpack::ValueType::String:
-            return left.stringRef().compare(right.stringRef());
-        default:
-            DCHECK(false) << "unsupport json type: " << (int)type;
+            case vpack::ValueType::String:
+                return left.stringRef().compare(right.stringRef());
+            default:
+                // other types like illegal, none, min, max are considered equal
+                return 0;
+            }
+        } else if (left.isInteger() && right.isInteger()) {
+            return left.getInt() - right.getInt();
+        } else {
+            return cmpDouble(left.getNumber<double>(), right.getNumber<double>());
         }
     } else {
+        if (left.type() == vpack::ValueType::MinKey) {
```

Comment:
left or right could both be minkey or maxkey?

Reply:
Yes. MinKey is not a key but a JSON type. JSON type of `vpack::ValueType` includes `INT, DOUBLE, STRING, NULL, NONE, MIN, MAX ...... `.